---
layout: post
author: "subin"
title:  "inferCNV"
subtitle: "Inferring copy number alterations from tumor single cell RNA-seq data"
type: "Tool"
category: "singleCell"
tags: ['inferCNV','copy number variation','scRNAseq','R']
disqus: true
---
:tada: 오늘은 월급날 얄루!

tumor sample로 scRNA seq data생산했을때, 가장 기본적으로 진행하는 분석이다. 어떤 cell이 cancer cell인지 알아내는것! 일반적인 cell type annotation tool(such as SingleR)을 사용해서 분석하면 cancer cell은 fibroblast나 epithelial cell같은 type으로 분류될 수 있다. <br/>
**inferCNV**는 RNAseq data에서 large-scale choromosomal Copy Number alteration을 추론해 normal에 비해 alteration이 많은 cancer cell을 구분한다. 

- [Installing of inferCNV](#installation)
- [Running inferCNV](#running-infercnv)
- [Results](#results)
<br/><br/><br/>

# Installation
## Pre-requisites
---
### JAGS (Just Another Gibbs Sampler)
install JAGS from source code (downlaod [here](https://sourceforge.net/projects/mcmc-jags/files/JAGS/4.x/Source/))
```bash
tar -xvzf JAGS-4.3.0.tar.gz
cd JAGS-4.3.0
./configure --prefix=[JAGS_installation_dir]
make & make install
```
<details>
<summary>:rotating_light: configure: error: "You need to install the LAPACK library"</summary>
<div markdown="1">
```bash
...
checking for cheev_ in -llapack_rs6k... no
configure: error: "You need to install the LAPACK library"
```
configuration하는 도중에 위와 같은 error가 나면, `lapack` 설치해준다 (download source from [here](https://github.com/Reference-LAPACK/lapack/archive/refs/tags/v3.9.1.tar.gz))
```bash
tar -xvzf v3.9.1.tar.gz
cd lapack-3.9.1
#make.inc.example 파일의 내용을 본인 환경에 맞게 수정한 후, make.inc파일로 저장한다.
mv make.inc.example make.inc 
make & make install
```
잘 설치되면 다시 한번 configuration시도해보고, lapack 깔았는데도 같은 에러가 난다면 이렇게 해보자
```
LDFLAGS="-L/source/lapack-3.9.1" F77=gfortran ./configure --prefix=/source/JAGS-4.3.0 --with-lapack=/source/lapack-3.9.1
```
</div></details>
## Install inferCNV
---
```R
if (!requireNamespace("BiocManager", quietly = TRUE))
     install.packages("BiocManager")
BiocManager::install("infercnv")
```


#Running inferCNV
## Input files
---
### Raw counts matrix (Gene x Cell)
> The counts matrix can be generated using any conventional single cell transcriptome quantification pipeline, yielding a matrix of genes (rows) vs. cells (columns) containing assigned read counts.

난 10X data사용하고 seurat으로 QC거친 data가 있으니까 거기에서 뽑아내면
```R
counts_matrix = GetAssayData(seurat_obj, slot="counts")
saveRDS(round(counts_matrix, digits=3), "sc.10x.counts.matrix")
```

### Sample Annotation file
>The sample annotation file is used to define the different cell types, and optionally, indicating how the cells should be grouped according to sample (ie. patient). The format is simply two columns, tab-delimited, and there is no column header.
>
>**Only those cells listed in the sample annotations file will be analyzed by inferCNV.**This is useful in case you cells of interest are a subset of the total counts matrix, without needing create a new matrix containing the subset of interest.

나는 immune cell type을 미리 나눠놨어서 그 cell들을 normal이라고 annotation 해서 넣을 예정
```R

```

### Gene ordering file
> The gene ordering file provides the chromosomal location for each gene. The format is tab-delimited and has no column header, simply providing the **gene name, chromosome, and gene span**

count할때 사용했던 reference와 같은 버전을 사용하자. (Gencode gtf file download [here](https://www.gencodegenes.org/mouse/release_M23.html))<br/>
bed file형식으로 만들기 위해 다음과 같은 과정을 거쳤다.
```bash
# gene_name  chromosome  start  end 의 포맷으로 만들기 위해 전처리
cat gencode.vM23.primary_assembly.annotation.gtf |  awk 'OFS="\t" {if ($3=="gene") {print $14,$1,$4-1,$5}}' | tr -d '";' > inferCNV_geneOrdering.txt
```
## InferCNV 2-step execution:
---
### Step1. Create inferCNV object

<details>
<summary>verbose</summary>
<div markdown="1">
```
INFO [2021-05-25 17:35:49] Parsing matrix: sc.10x.counts.matrix
INFO [2021-05-25 17:36:02] Parsing gene order file: gene_ordering_file.txt
INFO [2021-05-25 17:36:02] Parsing cell annotations file: cellAnnotations.txt
INFO [2021-05-25 17:36:02] ::order_reduce:Start.
INFO [2021-05-25 17:36:02] .order_reduce(): expr and order match.
INFO [2021-05-25 17:36:02] ::process_data:order_reduce:Reduction from positional data, new dimensions (r,c) = 16766,5684 Total=114172878 Min=0 Max=5041.
INFO [2021-05-25 17:36:02] num genes removed taking into account provided gene ordering list: 529 = 3.15519503757605% removed.
INFO [2021-05-25 17:36:02] -filtering out cells < 100 or > Inf, removing 0 % of cells
INFO [2021-05-25 17:36:04] validating infercnv_obj
```
</div></details>

### Step2. Create inferCNV object
